test/connection_switching_test.rb

```diff
@@ -579,6 +579,83 @@ def spec_name
     end
   end
 
+  describe "thread safety" do
+    Thread.abort_on_exception = true
+
+    describe "switching to from primary to replica in a thread" do
+      it "can safely from primary to replica" do
+        threads = []
+
+        threads << Thread.new { assert_using_primary_in_thread }
+        threads << Thread.new { assert_using_replica_in_thread }
+
+        wait_for_threads_to_be_ready(threads)
+        execute_and_wait_for_threads(threads)
+      end
+    end
+
+    it "can switch connections concurrently between shards, replica and primary in multiple threads" do
+      threads = []
+
+      threads << Thread.new { assert_using_shard_in_thread(0) }
+      threads << Thread.new { assert_using_shard_in_thread(1) }
+      threads << Thread.new { assert_using_replica_in_thread }
+      threads << Thread.new { assert_using_primary_in_thread }
+
+      wait_for_threads_to_be_ready(threads)
+      execute_and_wait_for_threads(threads)
+    end
+
+    def execute_and_wait_for_threads(threads)
+      threads.each { |t| t.wakeup if t.alive? }
+      threads.each(&:join)
+    end
+
+    def wait_for_threads_to_be_ready(threads)
+      sleep(0.01) until threads.all? { |t| t[:ready] }
+    end
+
+    def assert_using_shard_in_thread(shard_number)
+      ActiveRecord::Base.on_shard(shard_number) do
+        logger.debug "thread #{shard_number} sleep"
+        Thread.current[:ready] = true
+        sleep
+        assert_using_database("ars_test_shard#{shard_number}")
+        assert_using_database("ars_test_shard#{shard_number}", Ticket)
+        assert_using_database('ars_test', Account)
+        logger.debug "thread #{shard_number} done"
+      end
+    end
+
+    def assert_using_primary_in_thread
+      ActiveRecord::Base.establish_connection(::RAILS_ENV.to_sym)
```

Comment:
I'm not sure why this is required here for the primary but not for the shards or the replica.

Reply:
I don’t know either.

Reply:
It's because the test helper creates a new connection handler every time and this goes into a thread-isolated registry.

Spawned threads end up using the singleton `ActiveRecord::Base.default_connection_adapter` which is actually fine but the thread must `establish_connection`.